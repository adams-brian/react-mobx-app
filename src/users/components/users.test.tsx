import { mount, shallow } from 'enzyme';
import { Provider } from 'mobx-react';
import * as React from 'react';
import { MemoryRouter } from 'react-router';
import * as sinon from 'sinon';

import UsersState from '../state';
import EditUser from './editUser';
import UserList from './userList';
import WrappedUsers, { Users } from './users';

describe("Users", () => {

  let usersState: UsersState;
  let element: JSX.Element;

  beforeEach(() => {
    usersState = new UsersState({
      createUser: sinon.fake(),
      deleteUser: sinon.fake(),
      loadUsers: sinon.fake.resolves([]),
      updateUser: sinon.fake()
    });
    element = <Users usersState={usersState} />;
    return usersState.loadUsers();
  });

  it('renders as expected', () => {
    const component = shallow(element);
    expect(component).toMatchSnapshot();
  });
  
  it('routes /users to UserList', () => {
    const component = shallow(element);
    expect(component.find('Route[exact=true][path="/users"]').first().prop('component')).toBe(UserList);
  });

  it('routes /users/createuser to EditUser', () => {
    const component = shallow(element);
    expect(component.find('Route[exact=true][path="/users/createuser"]').first().prop('component')).toBe(EditUser);
  });

  it('routes /users/:id to EditUser', () => {
    const component = shallow(element);
    expect(component.find('Route[path="/users/:id"]').first().prop('component')).toBe(EditUser);
  });

});

describe("WrappedUsers", () => {

  let usersState: UsersState;
  let element: JSX.Element;

  beforeEach(() => {
    usersState = new UsersState({
      createUser: sinon.fake(),
      deleteUser: sinon.fake(),
      loadUsers: sinon.fake.resolves([]),
      updateUser: sinon.fake()
    });
    sinon.spy(usersState, 'loadUsers');
    element = (
      <MemoryRouter>
        <Provider usersState={usersState}>
          <WrappedUsers />
        </Provider>
      </MemoryRouter>
    );
  });

  it('renders as expected', () => {
    const component = mount(element);
    expect(component.find('Provider').first().children()).toMatchSnapshot();
  });

  it('calls loadUsers', () => {
    expect((usersState.loadUsers as sinon.SinonSpy).called).toBe(false);
    mount(element);
    expect((usersState.loadUsers as sinon.SinonSpy).calledOnce).toBe(true);
  });

});