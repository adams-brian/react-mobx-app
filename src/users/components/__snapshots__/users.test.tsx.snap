// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users renders as expected 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    path="/users"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/users/createuser"
  />
  <Route
    component={[Function]}
    path="/users/:id"
  />
</Switch>
`;

exports[`WrappedUsers renders as expected 1`] = `
<inject-loading(Users)-with-usersState>
  <loading(Users)
    usersState={
      UsersState {
        "api": Object {
          "createUser": [Function],
          "deleteUser": [Function],
          "loadUsers": [Function],
          "updateUser": [Function],
        },
        "users": Array [],
      }
    }
  />
</inject-loading(Users)-with-usersState>
`;
